### ISSUE TITLE
Comment on same line as code in script 

### ISSUE THREAD

**Issue by scopatz** [scopatz/xonsh#669]
The lexer / parser seems to hang indefinitely in the event that a comment follows a subprocess command in a script.  Namely, I found that the following failed:

```console
def func():
    ./setup.py install --user  # FIXME this should be made more general
```

**Comment by adqm** [comment]
Hmmm this doesn't seem to happen in all cases, though:
```
hartz@lappy-513 ~/x/x/parsers $ print(7) # hello
7
hartz@lappy-513 ~/x/x/parsers $ echo "hi" # hello
hi
hartz@lappy-513 ~/x/x/parsers $ ls -la #hello
total 108
drwxr-xr-x 3 hartz hartz  4096 Feb 11 08:27 .
drwxr-xr-x 5 hartz hartz  4096 Feb 11 08:27 ..
-rw-r--r-- 1 hartz hartz 80232 Feb 11 08:27 base.py
-rw-r--r-- 1 hartz hartz     0 Feb 11 08:27 __init__.py
drwxr-xr-x 2 root  root   4096 Feb 11 08:27 __pycache__
-rw-r--r-- 1 hartz hartz  5179 Feb 11 08:27 v34.py
-rw-r--r-- 1 hartz hartz  4937 Feb 11 08:27 v35.py
```

Seems to be subprocess commands _within functions_ that have comments after them.

I wonder if this is somehow getting caught in another loop of wrapping those lines with `$[...]` (maybe my fix from earlier doesn't catch all cases...it could be the whitespace at the start is confusing that fix).

**Comment by scopatz** [comment]
Ahhh interesting observation. We are probably not testing this particular case of function bodies.

**Comment by adqm** [comment]
In fact, it's not even an issue with that.  Any subproc-mode command with whitespace in front breaks.  I tracked it down to #601 being merged in where this started breaking (something with multiline not quite right), but I haven't yet dug deep enough to figure out exactly what change broke it.

Wrapping with `$[...]` is a workaround (whitespace before that structure is fine).

**Comment by adqm** [comment]
FTR, the test I used that broke things was: `<SPACE>echo "hi"<ENTER>`

According to my tests, that works on master before the multiline mode changes, and not after.

**Comment by scopatz** [comment]
> Wrapping with $[...] is a workaround (whitespace before that structure is fine).

Good to know.  Thanks for tracking down what caused this problem.

**Comment by gforsyth** [comment]
```
builtins.__xonsh_execer__.compile(' echo "hi"', mode='single', glbs=None, locs=builtins.__xonsh_ctx__)
```

hangs.  That's what's killing this in the multiline parser.  Investigating further...

**Comment by adqm** [comment]
Looks to me like the base reason is that the initial space is being treated as an `INDENT` token, and subproc mode doesn't know how to deal with that.

**Comment by gforsyth** [comment]
Ok.  I have some more thoughts/findings on this, but still no solution.  

For one thing, @scopatz original example will also hang the readline shell.

@adqm, your test which does break in PTK multiline works in the readline shell, because of `base_shell.precmd` which gets called by the readline shell and performs an `lstrip()`

oddly, the following hangs (since I guess it doesn't go through `precmd` for some reason?)

```Python
from xonsh.shell import Shell

rl = Shell(shell_type='readline')
rl.default(' echo "hi"')
```

All by way of saying, I think this goes deeper.  #601 is definitely where `<space>echo "hi"<enter>` started breaking on PTK, since that's when the `lstrip()` (via `precmd`) was removed, but the original failure persists going back to v0.2.4

**Comment by scopatz** [comment]
So, I think that `'  echo "hello"'` should be broken when passed directly into `default()`.  That is because this is the equivalent of started a Python module with a line that begins in whitespace.  That said, this should be a SyntaxError, not a hang.  

As I recall (and this was a while ago) @SmartViking and I had some discussion about this in #110 and #130.  The lstrip thing is a convenience for people at the shell and not part of the xonsh language.  precmd() is a shell notion, and not a language idea.

**Comment by scopatz** [comment]
Potential fix in #698
